File: backend/src/seeds/seed-items.ts
import fs from "fs";
import dotenv from "dotenv";
dotenv.config({ path: "../../.env" });
const { Client } = require("pg");

const client = new Client({
  host: process.env.POSTGRESQL_HOST,
  user: process.env.POSTGRESQL_USER,
  port: 5432,
  password: process.env.POSTGRESQL_PASSWORD,
  database: process.env.POSTGRESQL_DATABASE,
});

const updatedCurrency = JSON.parse(
  fs.readFileSync("../currencies/currency.json", "utf8")
);
const pictures: number[] = [];
for (let i = 0; i < 20; i++) {
  pictures.push(Math.ceil(Math.random() * 9007199254740990));
}

try {
  console.time("seed-time");
  await client.connect();
  await seedDB();
} catch (error) {
  console.error("Error:", error);
} finally {
  console.timeEnd("seed-time");
  console.log("Data seeding complete"); // [1913.70s] seed-time ~31 min
  client.end();
}

async function insertItem(row: any) {
  const universalType = Math.ceil(Math.random() * 6);
  let type = universalType;
  let euroPrice =
    universalType <= 3
      ? Math.ceil(Math.random() * 2000 + 400)
      : Math.ceil(Math.random() * 1000000 + 10000);

  const values = Object.values(updatedCurrency);
  const objectValue: number[] = values[
    Math.floor(Math.random() * values.length)
  ] as number[];
  const currencyCode = objectValue[0];
  const currencyName = objectValue[1];
  const currencySymbol = objectValue[2];
  let originalPrice = Math.round(euroPrice * objectValue[3]);
  while (originalPrice > 2000000000) {
    euroPrice =
      universalType <= 3
        ? Math.ceil(Math.random() * 2000 + 400)
        : Math.ceil(Math.random() * 1000000 + 10000);
    originalPrice = Math.round(euroPrice * objectValue[3]);
  }

  const city = row.city_name;
  const state = row.state_name;
  const country = row.country_name;
  const ranking = row.ranking;

  const coordinates = [
    row.lng + (ranking <= 2 ? Math.random() * 0.1 - 0.05 : 0),
    row.lat + (ranking <= 2 ? Math.random() * 0.1 - 0.05 : 0),
  ];
  if (coordinates[0] >= -32 && coordinates[0] < 62) {
    type += 6;
  } else if (coordinates[0] > 62) {
    type += 12;
  }

  const heating = Math.floor(Math.random() * 6); // 0: no, 1: yes, 2: oven, 3: heatpump, 4: boiler, 5: air
  const petsAllowed = Math.floor(Math.random() * 4); // 0: no pets, 1: yes, 2: only small pets, 3: no dogs
  const newItem = {
    currency_code: currencyCode,
    currency_name: currencyName,
    currency_symbol: currencySymbol,
    original_price: originalPrice,
    euro_price: euroPrice,
    country: country,
    state: state,
    city: city,
    coordinates: `POINT(${coordinates[0]} ${coordinates[1]})`,
    type: type,
    size: Math.ceil(Math.random() * 100 + 50),
    plot:
      universalType === 2 || universalType === 5
        ? Math.ceil(Math.random() * 2000) + 100
        : null,
    bed: Math.ceil(Math.random() * 4),
    bath: Math.ceil(Math.random() * 2),
    floors:
      universalType === 2 || universalType === 5
        ? Math.ceil(Math.random() * 3)
        : null,
    heating: heating,
    cooling: Math.random() > 0.5 ? true : null,
    utility_cost:
      universalType <= 3 ? `${Math.ceil(Math.random() * 250)}` : null,
    elevator:
      universalType === 2 || universalType === 5
        ? null
        : Math.random() > 0.5
        ? true
        : null,
    garden: Math.random() > 0.5 ? true : null,
    swimming_pool: Math.random() > 0.5 ? true : null,
    parking: Math.random() > 0.5 ? true : null,
    council_home: Math.random() > 0.5 ? true : null,
    is_deleted: Math.random() > 0.1 ? true : null,
    auction_date:
      Math.random() > 0.5 ? new Date().toISOString().split("T")[0] : null,
    pets_allowed: petsAllowed,
    realtor_fee_percentage: (Math.random() * 3 + 2).toFixed(2),
    deposit: Math.random() > 0.5 ? Math.ceil(Math.random() * 2000) : null,
    like_count: Math.ceil(Math.random() * 200),
    first_picture: pictures[0],
    pictures: pictures,
    user_id: "68bc150b-e522-4728-a1a8-120f2002207e",
    firm_id: null,
    firm_boost: 0,
    ranking: ranking,
  };

  const query = {
    text: `INSERT INTO items (currency_code, currency_name, currency_symbol, original_price, euro_price, country, state,
          city, coordinates, type, size, plot, bed, bath, floors, heating, cooling, utility_cost, elevator, garden,
          swimming_pool, parking, council_home, is_deleted, auction_date, pets_allowed, realtor_fee_percentage, deposit,
          like_count, first_picture, pictures, user_id, firm_id, firm_boost, ranking)
          VALUES ($1, $2, $3, $4, $5, $6, $7, $8, ST_GeogFromText($9), $10, $11, $12, $13, $14, $15, $16, $17, $18, $19, $20,
          $21, $22, $23, $24, $25, $26, $27, $28, $29, $30, $31, $32, $33, $34, $35)`,
    values: [
      newItem.currency_code,
      newItem.currency_name,
      newItem.currency_symbol,
      newItem.original_price,
      newItem.euro_price,
      newItem.country,
      newItem.state,
      newItem.city,
      `SRID=4326;POINT(${coordinates[0]} ${coordinates[1]})`,
      newItem.type,
      newItem.size,
      newItem.plot,
      newItem.bed,
      newItem.bath,
      newItem.floors,
      newItem.heating,
      newItem.cooling,
      newItem.utility_cost,
      newItem.elevator,
      newItem.garden,
      newItem.swimming_pool,
      newItem.parking,
      newItem.council_home,
      newItem.is_deleted,
      newItem.auction_date,
      newItem.pets_allowed,
      newItem.realtor_fee_percentage,
      newItem.deposit,
      newItem.like_count,
      newItem.first_picture,
      newItem.pictures,
      newItem.user_id,
      newItem.firm_id,
      newItem.firm_boost,
      newItem.ranking,
    ],
  };
  await client.query(query);
}

async function seedDB() {
  await client.query("DELETE FROM items");

  const resultCityCount = await client.query(`SELECT COUNT(*) FROM cities`);
  const cityCount = resultCityCount.rows[0].count;

  const offset = 100;
  let currentOffset = 0;

  const queryText = `
  SELECT *
  FROM cities
  LIMIT $1 OFFSET $2;`;

  for (let i = 0; i < cityCount / offset; i++) {
    const values = [offset, currentOffset];
    currentOffset += offset;
    const result = await client.query(queryText, values);
    console.log(i * offset);
    for (let j = 0; j < offset; j++) {
      const row = result.rows[j];
      let loopCount = 0;
      if (row?.ranking) loopCount = row.ranking <= 2 ? 10 : 1;
      for (let j = 0; j < loopCount; j++) {
        await insertItem(row);
      }
    }
  }
}

File: backend-go/database/seed_items.go
package main

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"time"

	"log"
	"math"
	"math/rand"
	"os"

	"github.com/Ali5937/fb-solid-start/backend-go/models"
	"github.com/Ali5937/fb-solid-start/backend-go/pgdb"
	"github.com/lib/pq"
)

func seedItems() {
	start := time.Now()
	currenciesData, keys := loadCurrencyData()
	db, err := pgdb.InitDB()
	if err != nil {
		log.Fatalf("Error opening database: %v", err)
	}
	defer db.Close()

	_, err = db.Exec("DELETE FROM items")
	if err != nil {
		log.Fatalf("Error deleting all rows from items table: %v", err)
	}

	var cityCount int
	err = db.QueryRow(`SELECT COUNT(*) FROM cities`).Scan(&cityCount)
	if err != nil {
		log.Fatalf("Error fetching row count of cities table: %v", err)
	}

	queryCitiesText := `
	SELECT *
	FROM cities
	LIMIT $1 OFFSET $2;`
	const offset = 100
	var currentOffset int = 0
	loopCount := cityCount / offset

	userIds := getUserIds(db)

	for i := 0; i <= loopCount; i++ {
		rows, err := db.Query(queryCitiesText, offset, currentOffset)
		if err != nil {
			log.Fatalf("Error fetching cities: %v", err)
		}
		defer rows.Close()
		currentOffset += offset
		userId := userIds[rand.Intn(len(userIds))]
		var pictures []int16
		for i := 0; i < 20; i++ {
			pic := int16(rand.Int31n(1000))
			pictures = append(pictures, pic)
		}

		var currentCity models.City
		for rows.Next() {
			err := rows.Scan(
				&currentCity.ID, &currentCity.City, &currentCity.State,
				&currentCity.Country, &currentCity.Ranking, &currentCity.Lat, &currentCity.Lng)
			if err != nil {
				log.Fatalf("Error reading row from rows in cities: %v", err)
			}
			insertItem(db, currenciesData, keys, currentCity, userId, pictures)
		}
		percentage := math.Round((float64(i)/float64(loopCount)*100)*100) / 100
		fmt.Printf("\r%v%%", percentage)
	}
	elapsed := time.Since(start).Milliseconds()
	fmt.Printf("\nTime taken seeding items: %vms", elapsed) //~62min-75min
}

func insertItem(db *sql.DB, currData map[string]models.Currency, ks []string, currentCity models.City,
	userId string, pictures []int16) {
	item := models.PostItem{}
	//1RentApartment, 2RentHouse, 3RentShared, 4BuyApartment, 5BuyHouse, 6BuyLand
	universalType := int16(rand.Int31n(6) + 1)

	var euroPrice int32
	if universalType <= 3 { //rent
		euroPrice = (rand.Int31n(2000) + 100)
		item.Floor = new(int16)
		*item.Floor = int16(rand.Int31n(10) - 1)
		item.UtilityCost = new(int16)
		*item.UtilityCost = int16(rand.Int31n(300))
		item.Deposit = new(int16)
		*item.Deposit = int16(rand.Int31n(euroPrice * (rand.Int31n(4) + 1)))
		item.CouncilHome = new(bool)
		*item.CouncilHome = rand.Int31n(2) == 0
		item.Plot = nil
		item.Floors = nil
	} else { //buy
		euroPrice = rand.Int31n(1000000) + 10000
		item.Plot = new(int32)
		*item.Plot = rand.Int31n(3000) + 100
		item.Floors = new(int16)
		*item.Floors = int16(rand.Int31n(3) + 1)
		item.Floor = nil
		item.UtilityCost = nil
		item.Deposit = nil
		item.CouncilHome = nil
	}

	randomCurrency := getRandomCurrency(currData, ks)
	originalPrice := math.Round(float64(euroPrice) * randomCurrency.ExchangeRate)

	item.EuroPrice = int32(euroPrice)
	item.OriginalPrice = int64(originalPrice)

	item.FirstPicture = int16(1)

	item.Type = universalType
	if item.Lng >= -32 && item.Lng < 62 {
		item.Type += 6
	} else if item.Lng > 62 {
		item.Type += 12
	}

	item.Size = int16(rand.Int31n(20000))
	item.Bed = int16(rand.Int31n(3) + 1)
	item.Bath = int16(rand.Int31n(2) + 1)
	item.LikeCount = int16(rand.Int31n(100))

	item.CurrencyCode = randomCurrency.Code
	item.CurrencyName = randomCurrency.Name
	item.CurrencySymbol = randomCurrency.Symbol
	item.FirmBoost = new(int16)
	*item.FirmBoost = 1
	item.PetsAllowed = new(bool)
	*item.PetsAllowed = rand.Int31n(2) == 0
	item.Ranking = currentCity.Ranking

	item.Country = currentCity.Country
	item.State = currentCity.State
	item.City = currentCity.City
	coordinates := fmt.Sprintf("SRID=4326;POINT(%v %v)", currentCity.Lng, currentCity.Lat)
	item.Heating = new(int16)
	*item.Heating = int16(rand.Int31n(6))
	item.IsDeleted = false
	item.Cooling = new(bool)
	*item.Cooling = rand.Int31n(2) == 0
	item.Elevator = new(bool)
	*item.Elevator = rand.Int31n(2) == 0
	item.Garden = new(bool)
	*item.Garden = rand.Int31n(2) == 0
	item.SwimmingPool = new(bool)
	*item.SwimmingPool = rand.Int31n(2) == 0
	item.Parking = new(bool)
	*item.Parking = rand.Int31n(2) == 0
	var isAuction = rand.Int31n(10) == 0
	if isAuction {
		item.AuctionDate = new(time.Time)
		*item.AuctionDate = time.Date(time.Now().Year(), time.Now().Month(), time.Now().Day(),
			0, 0, 0, 0, time.UTC)
	} else {
		item.AuctionDate = nil
	}

	tx, err := db.Begin()
	if err != nil {
		log.Fatalf("Error with transaction seeding items: %v", err)
	}
	defer tx.Rollback()

	queryText := `
  INSERT INTO items (user_id, original_price, euro_price, first_picture, type, size, plot, bed, bath, 
	like_count, firm_boost, pets_allowed, ranking, coordinates, city, state, country, currency_name, 
	currency_symbol, currency_code, floor, floors, heating, utility_cost, deposit, 
	is_deleted, cooling, elevator, garden, swimming_pool, parking, council_home, auction_date, firm_id)
  VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, ST_GeogFromText($14), $15, $16, $17,
	$18, $19, $20, $21, $22, $23, $24, $25, $26, $27, $28, $29, $30, $31, $32, $33, $34)
	RETURNING id`

	err = tx.QueryRow(queryText, userId, item.OriginalPrice, item.EuroPrice, item.FirstPicture, item.Type, item.Size,
		item.Plot, item.Bed, item.Bath,
		item.LikeCount, item.FirmBoost, item.PetsAllowed,
		item.Ranking, coordinates, item.City, item.State, item.Country, item.CurrencyName, item.CurrencySymbol,
		item.CurrencyCode, item.Floor, item.Floors,
		item.Heating, item.UtilityCost, item.Deposit,
		item.IsDeleted, item.Cooling, item.Elevator, item.Garden, item.SwimmingPool, item.Parking,
		item.CouncilHome, item.AuctionDate, item.FirmID).Scan(&item.Id)
	if err != nil {
		log.Fatalf("Error seeding items into db: %v", err)
	}

	tx.Exec(`
	INSERT INTO item_pictures (item_id, item_type, pictures)
	VALUES ($1, $2, $3)`, item.Id, item.Type, pq.Array(pictures))

	if err = tx.Commit(); err != nil {
		log.Fatalf("Error commiting transaction seeding items: %v", err)
	}

}

func loadCurrencyData() (map[string]models.Currency, []string) {
	file, err := os.Open("../currencies/currency-updated.json")
	if err != nil {
		log.Fatalf("Error opening currency-template.json: %v", err)
	}

	var currenciesData map[string]models.Currency
	var keys []string
	decoder := json.NewDecoder(file)
	err = decoder.Decode(&currenciesData)
	if err != nil {
		log.Fatalf("Error decoding JSON: %v", err)
	}

	keys = make([]string, 0, len(currenciesData))
	for key := range currenciesData {
		keys = append(keys, key)
	}

	return currenciesData, keys
}

func getRandomCurrency(currData map[string]models.Currency, ks []string) models.Currency {
	randomKey := rand.Int63n(int64(len(ks)))
	randomCurr := currData[ks[randomKey]]
	return randomCurr
}

func getUserIds(db *sql.DB) []string {
	var ids []string
	rows, err := db.Query(`
	SELECT id FROM users
	LIMIT 10`)
	if err != nil {
		log.Fatalf("Error fetching users: %v", err)
	}
	defer rows.Close()

	for rows.Next() {
		var id string
		err = rows.Scan(&id)
		if err != nil {
			log.Fatalf("Error getting id from user: %v", err)
		}
		ids = append(ids, id)
	}
	return ids
}
